### DISCUSSION TITLE
Support for I2C multiplexers multible busses

### DISCUSSION THREAD

**Discussion by plewka** [arendst/Tasmota#14028] [Ideas]
# Why/Idea:
In my hardware I'm using an I2C-Mux-Device (TCA9548A an old, barebone one). 
In software it only shows up as mask register which defines which secondary bus(es) is/are actually connected to the primary bus. Control/Driver is trivial...one transfered byte. I was now thinking about how to handle this with Tasmota with minimal changes... 

# How:
Since there are already two I2C interfaces in ESP32 and now supported, too: The idea would be to virtually define more I2C interfaces: First the direct ones and than the virtual ones. With my limited understanding I only see solvable issues like the MUX itself which must be prevented from being found more than once but I might be false. Is this the right way to go, or is there a better solution? Is there a chance the contribution is accepted if done right?

# Background:
I try to move my hardware - I did years ago - to Tasmota. Impossible to compete this great project. ;-) This is possible because of the ESP32 support and increased possibilities. Tasmota's room heat control has great features I didn't implement, too. But not all I implemented, though ;-). I'm very interested in that, too.
For multi room measurement I used I2C with low clock frequency (beside multichannel 1wire).
At I2C we managed to get as low as a few kHz which seems to be a lower limitation 
of the ESP32 or the sw stack. Fast signal falltime causes signal integrity issues, but can be handled.

**Comment by plewka** [comment]
## An idea to support multiple I2C buses for new drivers without forcing changes to old driver code? ## 
Isn't it possible to change the API with an intermediate "transparent" level which supports I2C device address as an uint16_t and use the address as two bytes -  one representing the bus and one representing the traditional address?
An old driver will simply access what it knows and only will look for devices on bus 0.
A newer driver can scan up to 256 additional buses for the specific device ID if it likes. It only needs a second loop
(may be with a limit define) in the driver to manipulate the upper byte. 
In the intermediate level you look for the bus number in the upper byte and select/switch the i2c infrastructure to the right bus before doing the access with the lower byte as the bus address. One thing which needs special service is to protect devices from being found more than once - especially the MUX devices. The same driver can't operate on the primary bus and on the secondary buses when using a MUX for sure. Could be fixed by blocking IDs, but shouldn't be relevant use case. 
Maybe this intermediate dispatching on different interfaces or MUXes can be done by using Berry since it is quite easy (at least in my limited understanding) and can be flexible enough to support different MUX ICs. Maybe the Berry code is triggered any time the upper byte of an I2C address is different to last access. This would basically limit overhead to one compare if this stuff is not active.
Maybe it's even possible to allow or suppress same address depending if the devices separate by their address or use the second I2C interface of the esp32. A simple implementation should be possible for esp8266, too.

Only an idea - let me know if I fully missed the point...

**Reply by s-hadinger** [reply]
This might be doable especially with esp32 and Berry. I will try to get a mux

**Comment by plewka** [comment]
I looked at TI so see the difference of the 10 i2C-Muxes they have. Basically it is all the same.
Some have more channels (max:8), some even control an IRQ mux/readback not required for tasmota, some have a reset some have more possible I2C addresses (all start at 0x70) and one has a car certification. The TCA, PCA difference is different voltage rules on the secondary bus, AFAIK.

They use priority encoding on rising channel, but I simply used a shifted '1'.
I don't need to switch back all secondary channels disabled.
If you like to have a look how I modified the BMP driver you may have a look at my fork.
I tried to do minimal changes and just corrected the id shown in sensor topic to take 
bus+id instead of i2c-address.
Look for the #ifdef TCA9548A_ADDR  which masks the changes. 
https://github.com/plewka/Tasmota/blob/development/tasmota/xsns_09_bmp.ino

**Comment by s-hadinger** [comment]
I received a `TCA9548A`. From what I understand it's more or less a switch to open connection with one of the 8 connectors. I.e. when you switch to 1 of 8 connectors, it's equivalent to this bus connected to host, and all 7 others disconnected.

This is unfortunately hard to map in a generic way to Tasmota. We could think of virtual I2C buses, or virtual IDs by using a prefix above the 8 low bits.

However there is no way to distinguish if a device is connected before the `TCA9548A` or if multiple devices with same ID are connected after the `TCA9548A`. This means that a basic mapping would make a single device appear 8 times, one on each virtual bus.

Any ideas?

**Reply by sfromis** [reply]
Simplest, but not great, would be to require that usage of TCA9548A precludes non-multiplexed devices on that physical I2C bus (of which there could be multiple).

Another strategy could be to require only using 7 connectors of the TCA9548A, meaning that the non-multiplexed devices (which cannot share I2C address with multiplexed ones) can be detected and flagged as such. They could be available on all 8 virtual busses, but it might make more sense to block them unless the blind channel is selected.

**Reply by rma-x** [reply]
The TCA9548A allows to detach all secondary busses at once, so only devices on the primary bus are visible, which can then be ignored when scanning the secondary busses.

Selection of the secondary busses is by bitmap, not by number, so any combination between none and all is possible, although selecting more than one bus at a time usually does not make much sense, but deselecting all does.

**Comment by plewka** [comment]
I had to change be response...
Not only the MUX itself must be blocked - any device found on higher level must 
be blocked for lower level to prevent it from getting mirrored. 
This sounds like it is useful to logically use levels...

You could go and split the additional 8 bits a second time and assign 2bits for 
the direct interface and 3 for the first level of MUXes and 3 for the second as a future 
enhancement.  
Since the higher level ID blocks the same on lower level you can't get full 4x8x8 of course.
ESP32 has got three hw I2C if I remember correctly which gives: 3x1x7 or 3x3x5 or 3x4x4...
You can even go further via sw-interfaces but...with Berry it should be possible
to give 3bits for the bus which would take 16(!) IOs...and go for 8x8.  

A structure which uses a MUX after a MUX should already be of quite limited use since
you even have to handle the electrical loads, too. A third level finally sounds "academic" to me.
The devices even start with all buses disabled where the first pull resistors have to come in. 
If you are able to connect 8 of these devices on top level you can build a quite big system - even 
with other I2C devices of low flexibility in address selection. 
For me the MUXes primary act as address collision savers than for connecting 
"as much as possible".

**Reply by plewka** [reply]
On closer look on permutations...there are max. 16 sub buses on one primary bus with current MUX devices from TI.

**Reply by rma-x** [reply]
If my math is right you can have a maximum of 64 sub busses on each primary bus without nesting/cascading, using eight TCA9548A devices each of which has eight sub busses. For proper operation this of course requires that at most one channel in one of the MUXes is selected at a time and all others are deselected.

I think this should be enough to not bother about cascading MUXes until somebody comes up who has an actual use case for it.
